seq2science/scripts/singlecell/sctk_qc.R

```diff
@@ -0,0 +1,253 @@
+qc_summary <- file.path(out_dir, "SCTK_CellQC_summary.csv", fsep = "/")
+qc_dir <- file.path(out_dir, "report", fsep = "/")
+pdf_out <- file.path(out_dir, "SCTK_DropletQC_figures.pdf", fsep = "/")
+numCores <- snakemake@threads
+
+# Log all console output
+log <- file(log_file, open = "wt")
+sink(log)
+sink(log, type = "message")
+
+# Log all variables for debugging purposes
+cat("# variables used for this analysis:\n")
+cat('log_file         <- "', log_file, '"\n', sep = "")
+cat('sample           <- "', sample, '"\n', sep = "")
+cat('replicates       <- "', replicates, '"\n', sep = "")
+cat('isvelo           <- "', isvelo, '"\n', sep = "")
+cat('rds_in           <- "', rds_in, '"\n', sep = "")
+cat('out_dir          <- "', out_dir, '"\n', sep = "")
+cat('rds_out          <- "', out_dir, '"\n', sep = "")
+cat('qc_dir           <- "', qc_dir, '"\n', sep = "")
+cat('qc_summary       <- "', qc_summary, '"\n', sep = "")
+cat('pdf_out          <- "', pdf_out, '"\n', sep = "")
+cat('data_type        <- "', data_type, '"\n', sep = "")
+cat('detect_mito      <- "', detect_mito, '"\n', sep = "")
+cat('mito_set         <- "', mito_set, '"\n', sep = "")
+cat('detect_cell      <- "', detect_cell, '"\n', sep = "")
+cat('cell_calling     <- "', cell_calling, '"\n', sep = "")
+cat('use_alt_exp      <- "', use_alt_exp, '"\n', sep = "")
+
+# Setup parallel type
+# https://github.com/compbiomed/singleCellTK/blob/master/exec/SCTK_runQC.R
+parallelType <- "MulticoreParam"
+Params <- list()
+
+if (numCores > 1) {
+  if (numCores > parallel::detectCores()) {
+    warning("numCores is greater than number of cores available. Set numCores as maximum number of cores available.")
+  }
+  numCores <- min(numCores, parallel::detectCores())
+  message(as.character(numCores), " cores are used for parallel computation.")
+
+  if (parallelType == "MulticoreParam") {
+    parallelParam <- MulticoreParam(workers = numCores)
+  } else if (parallelType == "SnowParam") {
+    parallelParam <- SnowParam(workers = numCores)
+  } else {
+    stop("'--parallelType' should be 'MulticoreParam' or 'SnowParam'.")
+  }
+  Params$QCMetrics$BPPARAM <- parallelParam
+  Params$emptyDrops$BPPARAM <- parallelParam
+  Params$doubletFinder$nCores <- numCores
+}
+
+# Modify sce object
+modifySCE <- function(sce) {
+  newnames <- gsub(x = rownames(sce), pattern = "\\..*$", replacement = "")
+  if (nrow(sce) == length(newnames)) {
+    rownames(sce) <- newnames
+  } else {
+    "Unequal gene sets: nrow(sce) != nrow(newnames)"
+  }
+  return(sce)
+}
+
+# Create scatter plot from alternative experiments
+plotAltExps <- function(out_dir, sce) {
+  pdf(file.path(out_dir, "SCTK_altexps.pdf", fsep = "/"))
+  for (n in altExpNames(sce)) {
+    x <- paste0("altexps_", n, "_percent")
+    y <- "detected"
+    print(plotColData(sce, x = x, y = y))
+  }
+  dev.off()
+}
+
+# Set sample col
+sample_col <- ifelse(replicates, "technical_replicates", "descriptive_name")
```

Comment:
ugh. I'm so sorry this is so convoluted in s2s 😭.

- `sample_col` is the "sample" column by default.
- If `"technical_replicates" %in% colnames(samples) & isTRUE(replicates)`, then it becomes "technical_replicates".
- However, if `"descriptive_name" %in% colnames(samples_sheet)`, then it becomes "descriptive_name" 
  - If the user added multiple descriptive names for the same technical replicate, the first descriptive name is used.

You've probably seen them before, but here is the [link to the relevant python code](https://github.com/vanheeringen-lab/seq2science/blob/1ab16b258d29159970edcba3235b89ffc840d880/seq2science/rules/merge_replicates.smk#L61).
And the [link to the R code](https://github.com/vanheeringen-lab/seq2science/blob/1ab16b258d29159970edcba3235b89ffc840d880/seq2science/scripts/deseq2/utils.R#L51) (used to replicate the Python `samples` in R).

Reply:
Thanks for the info, it is a bit clearer now :)

As it turns out, I do not need to set `sample_col` after all since all cells are from the same sample. One might only indicate the sample of origin in case of a merged or mixed sample (which does not apply here).